filogic: add support for Arcadyan WG620443

lekoOwO opened the pull request:
Inherit https://github.com/openwrt/openwrt/pull/17960 as the original PR has not been updated in a long time.

---
### Specification
Model: Arcadyan WG620443(-TC)
CPU: Mediatek MT7986a
RAM: 512MB DDR3
Flash: 128MB W25N01GVZE1G
WiFi: 2.4GHz 5GHz 4x4
GbE: 1 WAN and 1 LAN
Power: 12V 1A DC 5.5/2.1 Jack
LED: 3

### Installation for TC variant by CHT
#### Enable Telnet from stock WebUI
* Launch the browser to WebUI
* Navigate to `System` -`Remote Management`
* Launch Web DevTools then select the iframe of `remote_management.htm`
* Execute `closeDiv()` in the web console.
* Enable Telnet from LAN access, then click 'Save' button

#### Bypass the firmware validation
* Telnet into access point via LAN
* Login with default credential.
* In `Please enter No. or input command: ` prompt, type `Switch_To_SH` then Enter key
* In `Please input command!!>` prompt, type `cht_arc_JE1xxxxxxx` then Enter key
  
  * `JE1xxxxxxx` is the serial number of the access point, it shows in `System` -`Information`
* When shell popup, type `echo 1 /tmp/FW_check_dev.log` then Enter key to bypass firmware validation.

#### Install recovery image
* Navigate to `System` -`Firmware Update` in WebUI
* Select `openwrt-xx.xx.xx-initramfs-factory.bin` file
* Launch Web DevTools then select the iframe of `system_f.htm`
* Execute following JS script to starting flash the stage 1 firmware.

```
$("input[type=file]").attr("name","file")
var f=document.webForm;
f.action="upload.cgi";
f.ENCTYPE='multipart/form-data';
subForm({frm:f,genfrm:0,cmd:'',uploadtype:1,wizard:1,wait:2,done:function(){check_FW_and_jump();}});
```

* When the green LED light up,the recovery image is installed and booted

#### Install full image
* Launch browser to `http://192.168.1.1`
* Upgrade OpenWrt with `openwrt-xx.xx.xx-squashfs-sysupgrade.bin`
* Reboot

lekoOwO commented:
hi anyone please?

musashino205 commented on package/boot/uboot-tools/uboot-envtools/files/mediatek_filogic:
```diff
@@ -139,6 +139,10 @@ zyxel,ex5601-t0)
 zyxel,ex5700-telenor)
 	ubootenv_add_uci_config "/dev/ubootenv" "0x0" "0x4000" "0x4000" "1"
 	;;
+arcadyan,wg620443)
+	local envdev=/dev/mtd$(find_mtd_index "u-boot-env")
+	ubootenv_add_uci_config "$envdev" "0x0" "0x20000" "0x20000" "1"
+	;;
```
Please sort alphabetically.

lekoOwO replied:
done

musashino205 commented on target/linux/mediatek/filogic/base-files/etc/hotplug.d/ieee80211/11_fix_wifi_mac:
```diff
@@ -202,4 +202,9 @@ case "$board" in
 		[ "$PHYNBR" = "0" ] && macaddr_add $hw_mac_addr 1 > /sys${DEVPATH}/macaddress
 		[ "$PHYNBR" = "1" ] && macaddr_add $hw_mac_addr 2 > /sys${DEVPATH}/macaddress
 		;;
+	arcadyan,wg620443)
+		hw_mac_addr=$(mtd_get_mac_encrypted_arcadyan_mt7986 "pricfg" "89c72f31eb3243e8b0a781fe46a51c56" "87d81103921f4c17cbfd078b056f397a")
+		[ "$PHYNBR" = "0" ] && macaddr_add $hw_mac_addr 1 > /sys${DEVPATH}/macaddress
+		[ "$PHYNBR" = "1" ] && macaddr_add $hw_mac_addr 2 > /sys${DEVPATH}/macaddress
+		;;
```
ditto

musashino205 commented on target/linux/mediatek/filogic/base-files/lib/upgrade/platform.sh:
```diff
@@ -187,6 +187,10 @@ platform_do_upgrade() {
 		CI_ROOT_UBIPART=ubi
 		nand_do_upgrade "$1"
 		;;
+	arcadyan,wg620443)
+		fw_setenv boot_select 0
+		nand_do_upgrade $1
+		;;
```
ditto

lekoOwO replied:
done

musashino205 commented on target/linux/mediatek/filogic/base-files/lib/upgrade/platform.sh:
```diff
@@ -306,6 +310,14 @@ platform_pre_upgrade() {
 	xiaomi,mi-router-wr30u-stock|\
 	xiaomi,redmi-router-ax6000-stock)
 		xiaomi_initial_setup
+		;;
+	arcadyan,wg620443)
+		if [ "$(rootfs_type)" = "tmpfs" ]; then
+			fw_setenv bootargs console=ttyS0,115200n1 ubi.mtd=4 init=/sbin/init loglevel=8 earlycon=uart8250,mmio32,0x11002000
+			ubidetach -m 4 | true
+			ubiformat /dev/mtd4
+		fi
+		
```
ditto

lekoOwO replied:
done

musashino205 commented on target/linux/mediatek/image/filogic.mk:
```diff
@@ -2156,3 +2156,24 @@ define Device/zyxel_nwa50ax-pro
 endef
 TARGET_DEVICES += zyxel_nwa50ax-pro
+
+define Device/arcadyan_wg620443
```
ditto

lekoOwO replied:
done

musashino205 commented:
Please squash all commits to single one and follow the guideline.

https://openwrt.org/submitting-patches#submission_guidelines

lekoOwO committed:
mediatek: add support for Arcadyan WG620443(-TC)

Specification
-------------
- Model     : Arcadyan WG620443(-TC)
- CPU       : Mediatek MT7986a
- RAM       : 512MB DDR3
- Flash     : 128MB W25N01GVZE1G NAND Flash
- WLAN      : 2.4GHz 5GHz 4x4
- Ethernet  : 1x WAN, 1x LAN (Gigabit)
- Power     : 12V 1A DC (5.5/2.1 Jack, Center positive polarity)
- LED       : 3

Installation for TC variant by CHT
----------------------------------

1. Enable Telnet from stock WebUI:
   - Launch the browser to WebUI.
   - Navigate to `System` -> `Remote Management`.
   - Launch Web DevTools and select the iframe of `remote_management.htm`.
   - Execute `closeDiv()` in the web console.
   - Enable Telnet from LAN access, then click 'Save' button.

2. Bypass the firmware validation:
   - Telnet into access point via LAN.
   - Login with default credential.
   - In `Please enter No. or input command: ` prompt, type `Switch_To_SH` then Enter.
   - In `Please input command!!>` prompt, type `cht_arc_JE1xxxxxxx` then Enter.
     - `JE1xxxxxxx` is the serial number of the access point, found in `System` -> `Information`.
   - When shell popup, type `echo 1 /tmp/FW_check_dev.log` then Enter to bypass firmware validation.

3. Install recovery image:
   - Navigate to `System` -> `Firmware Update` in WebUI.
   - Select `openwrt-xx.xx.xx-initramfs-factory.bin` file.
   - Launch Web DevTools and select the iframe of `system_f.htm`.
   - Execute the following JS script to start flashing the stage 1 firmware:

     ```javascript
     $("input[type=file]").attr("name","file")
     var f=document.webForm;
     f.action="upload.cgi";
     f.ENCTYPE='multipart/form-data';
     subForm({frm:f,genfrm:0,cmd:'',uploadtype:1,wizard:1,wait:2,done:function(){check_FW_and_jump();}});
     ```

   - When the green LED lights up, the recovery image is installed and booted.

4. Install full image:
   - Launch browser to `http://192.168.1.1`.
   - Upgrade OpenWrt with `openwrt-xx.xx.xx-squashfs-sysupgrade.bin`.
   - Reboot.

Link: https://github.com/openwrt/openwrt/pull/19340

Co-authored-by: Power Li <pccr10001@gmail.com>
Signed-off-by: Leko Tseng <leko@leko.moe>

lekoOwO commented:
I have sort the lines, squashed the commits and rebased to the latest main branch

Please tell me if there's still anything wrong :)

musashino205 commented on target/linux/mediatek/dts/mt7986a-arcadyan-wg620443.dts:
```diff
@@ -0,0 +1,382 @@
+		wifi@0,0 {
+			compatible = "mediatek,mt76";
+			reg = <0x0000 0 0 0 0>;
+			mediatek,mtd-eeprom = <&factory 0xa0000>;
```
Please convert this to the NVMEM binding.

lekoOwO replied:
done

lekoOwO commented on target/linux/mediatek/dts/mt7986a-arcadyan-wg620443.dts:
```diff
@@ -0,0 +1,382 @@
+	pinctrl-names = "default";
+	pinctrl-0 = <&wf_2g_5g_pins>;
+
+	mediatek,mtd-eeprom = <&factory 0x0>;
```
done

musashino205 commented on target/linux/mediatek/dts/mt7986a-arcadyan-wg620443.dts:
```diff
@@ -0,0 +1,392 @@
+				reg = <0x180000 0x200000>;
+				read-only;
+
+				compatible = "nvmem-cells";
```
Please update this to the nvmem-layout binding. (`compatible = "fixed-layout"`)

lekoOwO replied:
done :)

abajk commented on target/linux/mediatek/dts/mt7986a-arcadyan-wg620443.dts:
```diff
@@ -0,0 +1,394 @@
+		mediatek,bmt-max-ratio = <1>;
+		mediatek,bmt-max-reserved-blocks = <64>;
+
+		spi-max-frequency = <20000000>;
```
On most boards 52MHz works fine. Can it go higher?

lekoOwO replied:
hmmm idk, didn't try it though
my devices are all in prod now working fine :)

abajk replied:
With a higher frequency, the device should boot faster, and so on. It would be good to use the same frequency across all boards. Take a look at this [MR](https://github.com/openwrt/openwrt/pull/18752) :)

hauke commented on target/linux/mediatek/dts/mt7986a-arcadyan-wg620443.dts:
```diff
@@ -0,0 +1,394 @@
+		compatible = "gpio-leds";
+
+		led_status_red: red {
+			label = "red:net";
```
Please use:
```
color = <LED_COLOR_ID_RED>;
function: <LED_FUNCTION_LAN>;
```
and remove the `label` property. This will generate this label: `red:lan`

Please do the same for the other LEDs.

Rondom commented on package/boot/uboot-tools/uboot-envtools/files/mediatek_filogic:
```diff
@@ -57,6 +57,10 @@ huasifei,wh3000|\
 	;;
+arcadyan,wg620443)
+	local envdev=/dev/mtd$(find_mtd_index "u-boot-env")
+	ubootenv_add_uci_config "$envdev" "0x0" "0x20000" "0x20000" "1"
```
I think we can use ubootenv_add_mtd instead and keep referencing by name.
